dali/kernels/common/join/tensor_join_gpu_impl.cuh

```diff
@@ -0,0 +1,154 @@
+      lo = m + 1;
+    else
+      break;  // Found! For stacking this will always work unless precision fails
+    m = (lo + hi) >> 1;  // proceed to regular binary search
+  } while (lo <= hi);
+  return m;
+}
+
+template <typename Element>
+__device__ void JoinTensors(OutputDesc<Element> out,
+                            const InputDesc<Element> *__restrict__ in,
+                            int njoin) {
+  uint64_t block_size = static_cast<uint64_t>(blockDim.x);
+  uint64_t out_offset = threadIdx.x + block_size * blockIdx.x;
+  uint64_t step = block_size * gridDim.x;
+  for (; out_offset < out.total_size; out_offset += step) {
+    uint64_t join_offset;
+    uint64_t outer = div_mod(join_offset, out_offset, out.outer_stride);
+    int t = FindTensor(join_offset, out.guess_tensor_mul, in, njoin);
+    ptrdiff_t offset_in_tensor = join_offset - in[t].join_offset;
+    out.data[out_offset] = in[t].data[offset_in_tensor + outer * in[t].outer_stride];
+  }
+}
+
+template <typename Element>
+__global__ void JoinTensorsKernel(const OutputDesc<Element> *__restrict__ out,
+                                  const InputDesc<Element> *__restrict__ in,
+                                  int njoin) {
+  int sample_idx = blockIdx.y;
+  JoinTensors(out[sample_idx], in + sample_idx * njoin, njoin);
+}
+
+
+/**
+ * @brief Populates the input and output descriptors given the input and output tensor lists.
+ *
+ * @tparam ElementType tensor element
+ * @tparam out_ndim   dimensionality of the output, typically DynamicDimensions
+ * @tparam in_ndim    dimensionality of the inputs, typically DynamicDimensions
+ *
+ * @remarks
+ * The descirptors are the same when concatenating and stacking, hence no `new_axis` parameter.
+ */
+template <typename ElementType, int out_ndim, int in_ndim>
+void FillDescs(span<OutputDesc<ElementType>> output_descs,
+               span<InputDesc<ElementType>> input_descs,
+               const OutListGPU<ElementType, out_ndim> &output,
+               span<const InListGPU<ElementType, in_ndim> *const> inputs,
+               int axis) {
+  int njoin = inputs.size();
+  int N = output.num_samples();
+  assert(static_cast<int>(output_descs.size()) == N);
+  assert(static_cast<int>(input_descs.size()) == N * njoin);
+
+  for (int i = 0; i < N; i++) {
+    int64_t join_offset = 0;
+    for (int t = 0; t < njoin; t++) {
+      auto in_shape = inputs[t]->tensor_shape_span(i);
+      auto &desc = input_descs[i*njoin+t];
+      desc.data = inputs[t]->data[i];
+      desc.outer_stride = volume(in_shape.begin() + axis, in_shape.end());
+      desc.join_offset = join_offset;
+      join_offset += desc.outer_stride;
+    }
+
+    auto out_shape = output.tensor_shape_span(i);
+    auto &out_desc = output_descs[i];
+    out_desc.data = output.data[i];
+    auto join_size = volume(out_shape.begin() + axis, out_shape.end());
+    out_desc.outer_stride = join_size;
+    out_desc.total_size = volume(out_shape);
+    if (njoin == 2) {
```

Comment:
Do we enforce njoin > 1? Maybe we should?

Reply:
No, we don't. And it does no harm if njoin is 1, however stupid it might be. The operator, however, has a fast path for njoin==1 (just copies).

Reply:
ok